rainier-core/src/main/scala/com/stripe/rainier/compute/Encoder.scala

```diff
@@ -0,0 +1,60 @@
+package com.stripe.rainier.compute
+
+trait Encoder[T, U] {
+  def wrap(t: T): U
+  def create(acc: List[Variable]): (U, List[Variable])
+  def extract(t: T, acc: List[Double]): List[Double]
+}
+
+object Encoder {
+  implicit val int: Encoder[Int, Real] =
+    new Encoder[Int, Real] {
+      def wrap(t: Int) = Real(t)
+      def create(acc: List[Variable]) = {
+        val u = new Variable
+        (u, u :: acc)
+      }
+      def extract(t: Int, acc: List[Double]) =
+        t.toDouble :: acc
+    }
+
+  implicit val double: Encoder[Double, Real] =
+    new Encoder[Double, Real] {
+      def wrap(t: Double) = Real(t)
+      def create(acc: List[Variable]) = {
+        val u = new Variable
+        (u, u :: acc)
+      }
+      def extract(t: Double, acc: List[Double]) =
+        t :: acc
+    }
+
+  implicit def zip[A, B, X, Y](implicit a: Encoder[A, X],
+                               b: Encoder[B, Y]): Encoder[(A, B), (X, Y)] =
+    new Encoder[(A, B), (X, Y)] {
+      def wrap(t: (A, B)) = (a.wrap(t._1), b.wrap(t._2))
+      def create(acc: List[Variable]) = {
+        val (bv, acc1) = b.create(acc)
+        val (av, acc2) = a.create(acc1)
+        ((av, bv), acc2)
+      }
+      def extract(t: (A, B), acc: List[Double]) =
+        a.extract(t._1, b.extract(t._2, acc))
+    }
+
+  def vector[T, U](size: Int)(
```

Comment:
Should this be an `implicit def vector`?

Reply:
It can't be because it can't implicitly know the size, so you have to tell it that.

We could also/alternatively do a `ToVector` kinda like your `ToMap` to solve this. So eg maybe we could have you do:

 ```scala
case class MyDataVec(vec: Seq[Double])
object MyDataVec {
 implicit val toVec = new ToVector[MyDataVec] {
    val size = 10
    def apply(v: MyDataVec) = v.vec
}
```

and then we could have an implicit for that.